### ISSUE TITLE
Replace spaces or URLEscape screenshot/html filenames

### ISSUE THREAD

**Issue by treyturner** [AOEpeople/geb-spock-reports#25]
Hi there,
Would you consider having geb-spock-reports replace spaces in the PNG and HTML filenames with an underscore (or otherwise URL-escaping their filenames)? I'm attempting to publish my reports to Gitlab, but it can't handle spaces in these filenames, so the links end up broken. If you don't want to do this for everyone by default, it could still be useful to have this feature via configuration.

Thanks for your time and work on the reporter.

**Comment by tilmanginzel** [comment]
Hey @treyturner, I just checked and this seems to be a bit more complicated. I'll try to explain why.

The file name is generated by Geb itself, specifically [here](https://github.com/geb/geb/blob/master/module/geb-spock/src/main/groovy/geb/spock/GebReportingSpec.groovy#L59) and [here](https://github.com/geb/geb/blob/master/module/geb-core/src/main/groovy/geb/report/ReporterSupport.groovy#L26). As you can see, the filename simply consists of some indices plus the _feature name_ plus the _label_ of the `report` method . This label is then used in my [GebReportingListener](https://github.com/AOEpeople/geb-spock-reports/blob/master/src/main/groovy/com/aoe/gebspockreports/GebReportingListener.groovy#L46) [1]. Now, you _could_ extend the default `GebReportingSpec` and overwrite the `createReportLabel` method with a custom escaped string. So far straightforward.

The complicated part is now in the template. As you might know, this library mostly consists of a custom template for [spock-reports](https://github.com/renatoathaydes/spock-reports). There is some admittedly [weird](https://github.com/AOEpeople/geb-spock-reports/blob/master/src/main/resources/templates/spec-template.html#L162) [logic](https://github.com/AOEpeople/geb-spock-reports/blob/master/src/main/groovy/com/aoe/gebspockreports/report/SpecReport.groovy#L40) in my template to find the correct Geb artifacts in the json file by the _feature name_. This feature name now comes from spock-reports. So even if you escape the label in the file names, my template then cannot find the files as I am searching them by an index and the original feature name which is provided by spock-reports (unescaped, with spaces).

**Possible solution 1**
I currently do not see a good reason why I am trying to find the Geb artifacts by the _feature name_. Maybe I could drop that condition and simple search by the index. In that case, writing a custom GebReportingSpec with an escaped label could be sufficient.

**Possible solution 2**
I could rename/escape all files right at the beginning of the report creation. Some further logic would need to be adjusted in the template, but sounds doable. 

Solution 2 sounds more bulletproof and would also benefit others using Gitlab, so I probably would prefer that one. Will need to check when I can find the time for it though. :)

[1] The GebReportingListener basically writes a json file with all relevant info, so that once the report is being generated, all the data (file paths etc) can be fetched from this json file.

**Comment by treyturner** [comment]
That all makes sense, thanks for the detailed response!

**Comment by tilmanginzel** [comment]
Hi @treyturner, I finally found some time to look into this. A fix is implemented in #26. Before merging, I want to test it on a larger code base in the coming days. I have attached a [current snapshot](https://github.com/AOEpeople/geb-spock-reports/files/4399138/geb-spock-reports-0.2.6-SNAPSHOT.jar.zip) in case you want to test this before I publish it.

The escaping can be enabled by a config flag, but by default it's enabled so nothing to be done from your part. :)
